src/Storages/MergeTree/MergeTreeTransform.cpp

```diff
@@ -0,0 +1,134 @@
+#include <Storages/MergeTree/MergeTreeTransform.h>
+#include <Storages/MergeTree/MergeTreeBlockReadUtils.h>
+#include <Storages/MergeTree/LoadedMergeTreeDataPartInfoForReader.h>
+#include <Storages/MergeTree/IMergeTreeReader.h>
+#include <Columns/ColumnLazy.h>
+#include "base/defines.h"
+
+namespace DB
+{
+
+Block MergeTreeTransform::transformHeader(Block header)
+{
+    for (auto & it : header)
+    {
+        if (isColumnLazy(*(it.column)))
+        {
+            it.column =  it.type->createColumn();
+        }
+    }
+    return header;
+}
+
+MergeTreeTransform::MergeTreeTransform(
+    const Block & header_,
+    const MergeTreeData & storage_,
+    const StorageSnapshotPtr & storage_snapshot_,
+    const LazilyReadInfoPtr & lazily_read_info_,
+    const ContextPtr & context_,
+    const AliasToNamePtr & alias_index_)
+    : ISimpleTransform(
+    header_,
+    transformHeader(header_),
+    true)
+    , storage(storage_)
+    , storage_snapshot(storage_snapshot_)
+    , use_uncompressed_cache(context_->getSettings().use_uncompressed_cache)
+    , data_parts_info(std::move(lazily_read_info_->data_parts_info))
+{
+    for (const auto & it : header_)
+    {
+        if (isColumnLazy(*(it.column)))
+        {
+            requested_column_names.emplace_back((*alias_index_)[it.name]);
+            alias_column_names.emplace_back(it.name);
+        }
+    }
+}
+
+void MergeTreeTransform::transform(Chunk & chunk)
+{
+    const size_t columns_size = alias_column_names.size();
+    const size_t rows_size = chunk.getNumRows();
+    auto columns = chunk.detachColumns();
+
+    Block block = getInputPort().getHeader().cloneWithColumns(columns);
+    const auto * column_lazy = typeid_cast<const ColumnLazy *>(block.getByName(alias_column_names[0]).column.get());
+    const ColumnUInt64 * part_num_column = typeid_cast<const ColumnUInt64 *>(&column_lazy->getPartNumsColumn());
+    const ColumnUInt64 * row_num_column = typeid_cast<const ColumnUInt64 *>(&column_lazy->getRowNumsColumn());
+
+    ReadSettings read_settings;
+    read_settings.direct_io_threshold = 1;
+    MergeTreeReaderSettings reader_settings =
+    {
+        .read_settings = read_settings,
+        .save_marks_in_cache = false,
+    };
+
+    MutableColumns lazily_read_columns;
+    lazily_read_columns.resize(columns_size);
+    for (size_t i = 0; i < alias_column_names.size(); ++i)
+    {
+        const auto & alias_name = alias_column_names[i];
+        lazily_read_columns[i] = block.getByName(alias_name).type->createColumn();
+        lazily_read_columns[i]->reserve(rows_size);
+    }
+
+    for (size_t row_idx = 0; row_idx < rows_size; ++row_idx)
```

Comment:
This is quite expensive. Although we cannot directly read one row from a granule, at least we should avoid deserializing unnecessary rows. I think we need to build some infrastructure methods in `IMergeTreeReader`.